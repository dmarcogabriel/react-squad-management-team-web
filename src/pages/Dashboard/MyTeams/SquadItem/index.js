import React from 'react';
import { MdShare, MdEdit } from 'react-icons/md';
import { FaTrashAlt } from 'react-icons/fa';
import cn from 'classnames';
import { Link } from 'react-router-dom';

const SquadItem = ({
  squad,
  onClick,
  onDelete,
  dataTestid,
  squadSelectButtonTestid,
  squadDeleteButtonTestid,
}) => (
  <tr>
    <td>
      <div
        data-testid={squadSelectButtonTestid || 'squadItem'}
        role="button"
        tabIndex={0}
        onClick={() => onClick(squad)}
        onKeyDown={(e) => e.keyCode === 13 && onClick(squad)}
        className={cn(
          'relative border-b-2 border-gray-200 rounded p-2 flex w-full',
          squad.selected && 'text-primary-dark bg-primary-light'
        )}
      >
        <p data-testid={dataTestid} className="inline mr-10">
          {squad.name}
        </p>

        <p className="inline ml-2">{squad.description.substring(0, 20)}</p>

        <div
          data-testid="actionButtons"
          className={cn(
            'flex items-center absolute h-full',
            !squad.selected && 'hidden',
            'right-0 top-0 mr-3'
          )}
        >
          <button
            data-testid={squadDeleteButtonTestid || 'squadItem-delete'}
            type="button"
            onClick={() => onDelete(squad.id)}
          >
            <FaTrashAlt />
          </button>

          <button type="button" className="mx-2">
            <MdShare />
          </button>

          <Link to={`create?squadId=${squad.id}`}>
            <MdEdit />
          </Link>
        </div>
      </div>
    </td>
  </tr>
);

export default SquadItem;
