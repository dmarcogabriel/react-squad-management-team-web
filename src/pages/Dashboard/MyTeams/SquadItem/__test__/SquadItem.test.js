import { render, fireEvent } from '@testing-library/react';
import React from 'react';
import ReactDOM from 'react-dom';
import { BrowserRouter as Router } from 'react-router-dom';
import renderer from 'react-test-renderer';
import SquadItem from '..';

const mockSquadSelected = {
  id: 'm1',
  name: 'mock1',
  description: 'This is the description of the squad.',
  selected: true,
};

const mockSquadNotSelected = {
  id: 'm2',
  name: 'mock2',
  description: 'This is the description of the squad.',
  selected: false,
};

const Component = (props) => (
  <Router>
    <SquadItem {...props} />
  </Router>
);

describe('<SquadItem />', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');

    ReactDOM.render(<Component squad={mockSquadNotSelected} />, div);
  });

  it('renders not selected classes correctly', () => {
    const { getByTestId } = render(<Component squad={mockSquadNotSelected} />);

    expect(getByTestId('squadItem')).toHaveClass(
      'relative border-b-2 border-gray-200 rounded p-2 flex w-full'
    );
  });

  it('renders selected classes correctly', () => {
    const { getByTestId } = render(<Component squad={mockSquadSelected} />);

    expect(getByTestId('squadItem')).toHaveClass(
      'relative border-b-2 border-gray-200 rounded p-2 flex w-full text-primary-dark bg-primary-light'
    );
  });

  it('execute select squad event', () => {
    let result;

    const mockFunction = (selectedSquad) => {
      result = selectedSquad;
    };

    const { getByTestId } = render(
      <Component onClick={mockFunction} squad={mockSquadNotSelected} />
    );

    const squadItem = getByTestId('squadItem');
    fireEvent.click(squadItem);

    expect(result).toEqual(mockSquadNotSelected);
  });

  it('execute onDelete event', () => {
    const mock = {
      onDelete: (squadId) => squadId,
      onClick: () => {},
    };
    const spyDelete = jest.spyOn(mock, 'onDelete');
    const spySelect = jest.spyOn(mock, 'onClick');

    const { getByTestId } = render(
      <Component
        onDelete={spyDelete}
        onClick={spySelect}
        squad={mockSquadSelected}
      />
    );

    const squadItemDeleteButton = getByTestId('squadItem-delete');
    fireEvent.click(squadItemDeleteButton);

    expect(spyDelete).toHaveBeenCalled();
  });

  it('matches snapshot', () => {
    expect(
      renderer.create(<Component squad={mockSquadSelected} />).toJSON()
    ).toMatchSnapshot();
  });
});
