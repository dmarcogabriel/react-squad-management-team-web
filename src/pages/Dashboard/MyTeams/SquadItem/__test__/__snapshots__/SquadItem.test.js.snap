// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SquadItem /> matches snapshot 1`] = `
<tr>
  <td>
    <button
      className="relative border-b-2 border-gray-200 rounded p-2 flex w-full text-primary-dark bg-primary-light"
      data-testid="squadItem"
      onClick={[Function]}
      type="button"
    >
      <p
        className="inline mr-10"
      >
        mock1
      </p>
      <p
        className="inline ml-2"
      >
        This is the descript
      </p>
      <div
        className="flex items-center absolute h-full right-0 top-0 mr-3"
        data-testid="actionButtons"
      >
        <button
          data-testid="squadItem-delete"
          onClick={[Function]}
          type="button"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M32 464a48 48 0 0 0 48 48h288a48 48 0 0 0 48-48V128H32zm272-256a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zM432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16z"
            />
          </svg>
        </button>
        <button
          className="mx-2"
          type="button"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7s-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81 1.66 0 3-1.34 3-3s-1.34-3-3-3-3 1.34-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9c-1.66 0-3 1.34-3 3s1.34 3 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.16c-.05.21-.08.43-.08.65 0 1.61 1.31 2.92 2.92 2.92 1.61 0 2.92-1.31 2.92-2.92s-1.31-2.92-2.92-2.92z"
            />
          </svg>
        </button>
        <a
          href="/create?squadId=m1"
          onClick={[Function]}
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34c-.39-.39-1.02-.39-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
            />
          </svg>
        </a>
      </div>
    </button>
  </td>
</tr>
`;
